### PR TITLE
Remove `_pipeline` from linker and refactor CTE pipeline

### PR DESCRIPTION
**Pull request by RobinL**
This PR:
- eliminates the shared `linker._pipeline`, so that all SQL operations create and use fresh pipeline(s). 
- eliminated associated linker methods `linker._enqueue_sql`. `linker._execute_sql_pipeline`
- **all** SQL execution now happens via a pipeline that gets passed to [database_api.sql_pipeline_to_splink_dataframe](https://github.com/moj-analytical-services/splink/blob/16a9e93feec93ff138a06539e7f8fd089490c208/splink/database_api.py#L181)
- eliminates confusing linker methods `_initialise_df_concat` and `_initialise_df_concat_with_tf`, instead relying on more explicit calculations using functions in `vertically_concatenate.py`
- clarifies the naming of pipeline classes to be more precise (Task is now CTE, and SQLPipeline is now CTEPipeline)
- removes the option of passing `input_dataframes` to `sql_pipeline_to_splink_dataframe`, instead, input dataframes can be added to a `CTEPipeline` with `.append_input_dataframe`.  Users can therefore add dataframes in places that make the logic flow more clearly
- Pipelines are now 'single use'. Once executed, they will not be allowed to be re-used; a fresh one has to be created.  A `pipeline.spent` property enforces this.

## Motivation for this PR

Consider the existing [_initialise_df_concat_with_tf](https://github.com/moj-analytical-services/splink/blob/d41c3a0d9fd1753678c5d6a9afc2185587b8756a/splink/linker.py#L507).

The return type and the mutations of state it performs are confusing:
- If you set `materialise=True` it returns a Splink dataframe.   
- If you set `materialise=False`, it enqueues sql on `linker._pipeline()` and returns None

So it relies on:
- the ability to mutate `linker._pipeline`
- The ability to execute `linker._pipeline`, and reuse it

This function is not really compatible with the idea of using a fresh SQL pipeline each time we want to queue sql.  You'd need to pass a pipeline in, but it's not clear what comes out.

By allowing input tables to be queued directly onto the `CTEPipeline`, then we can write a new function `linker._enqueue_df_concat_with_tf` which:
- If you set `materialise=True`, runs the SQL, and returns a `CTEPipeline` with the result already enqueued
- If you set `materialise=False`, enqueues the SQL without running it to the pipeline, and returns a CTEPipeine 

Thus allowing us to replace all use of `_initialise_df_concat_with_tf`

Also closes https://github.com/moj-analytical-services/splink/issues/1696

## Reviewing

The main changes have been on:
- pipeline.py
- linker.py
- database_api.py

All the other changes are all just downstream consequences of changing those files

I was getting weirdness with the tests so you'll see i had to reset the cache

## Related possible future PRs

- Should the pipeline have a `pipeline.to_splink_dataframe(db_api)` method?  This feels clearer than having to call `db_api.sql_pipeline_to_splink_dataframe(pipeline)`

### CODE REVIEW

**Comment by RobinL** [comment]
TODO next:
Find new records:
```
rec = df.to_dict(orient="records")[0]
rec
linker.find_matches_to_new_records([rec]).as_pandas_dataframe()
df_predict.as_pandas_dataframe()
```

**Comment by RobinL** [comment]
TODO next:
Get rid of _initialise_df_concat

File: tests/test_caching.py [hunk 1]
```diff
@@ -51,45 +58,6 @@ def test_cache_id(tmp_path):
     assert linker_uid == random_uid
 
 
-def test_materialising_works():
```
Comment (RobinL) [inline tests/test_caching.py:54]: I don't think these tests are needed now we're explicitly using a fresh pipeline for everything - there's no longer the potential for accidentally forgetting to reset the pipeline etc. and any future errors should be much clearer if they do occur

**Comment by RobinL** [comment]
Next need to fix `_sql_to_splink_dataframe_checking_cache`

**Comment by RobinL** [comment]
There are some references to `_execute_sql_pipeline` left in the tests

**Comment by RobinL** [comment]
Next:
- need to restore `linker_table_to_splink_dataframe`
- When pipeline gets used by `sql_pipeline_to_splink_dataframe`, set spent to true.  Remove 'reusable' option. i.e. ensure no pipelines are reused

**Review by ADBond — APPROVED** [review]
I think this is great, looks like a big QoL improvement! Feels a lot clearer treating input frames in this way also - used to often get a bit tripped up by the old way, as it didn't line up between when you need/use them.

Haven't looked in detail at everything, but happy that the shape of this is good, and sure we can pick up any small issues if any arise

**Pull request merged**